### PR TITLE
Add traceback module

### PR DESCRIPTION
**Pull request by microDev1**
This PR implements `CPython` compatible `traceback` module.

### CODE REVIEW

**Comment by microDev1** [comment]
- Only `print_exception()` is currently implemented. I was considering implementing `format_exception()` (only difference is that this returns a string) but ran into trouble with compressed strings `%q` specifier. I am thinking of adding `mp_sprintf()` but haven't made any progress on that front yet.
- Following args are ignored for the reasons given and are only implemented to maintain compatibility with `CPython`.
  - `etype` - From `Python-3.5` this is ignored and is inferred from type of value internally.
  - `tb` - The `__traceback__` object is not exposed at the moment and is obtained internally; however, if set to `None`, traceback will not be printed, similar to `CPython`.
  - `chain` - Chained exceptions are not yet implemented.

**Comment by tannewt** [comment]
>   * `tb` - The `__traceback__` object is not exposed and is obtained internally;
>       however, if this is set to `None`, traceback will not be printed, similar to `CPython`.

Is this parameter required in CPython? What will we provide in CircuitPython? Does it do the same thing in CPython? Would you mind providingn an example use?

**Comment by microDev1** [comment]
> Is this parameter required in CPython?

Yes, this is a required parameter.

> What will we provide in CircuitPython? 

As this is ignored, anything (i.e. `Type: Any`) can be supplied to the function call.

> Does it do the same thing in CPython?

Yes, the current implementation is same as in `CPython`. The only difference is:

- `CPython` - The traceback is obtained from the supplied traceback object.
- `CircuitPython` - There is no traceback object (i.e. `Type: TracebackType`) implemented so this is obtained internally.

Only thing I can think of that won't be possible because of the above is supplying traceback from a different exception.

> Would you mind providing an example use?

```python
try:
    raise Exception("xyz")
except Exception as exc:
    # CPython:
    traceback.print_exception(None, exc, None)
    traceback.print_exception(None, exc, exc.__traceback__)
    traceback.print_exception(None, exc, exc.__traceback__, limit=1, file=log, chain=True)
    # CircuitPython:
    traceback.print_exception(None, exc, None)
    traceback.print_exception(None, exc, "any")
    traceback.print_exception(None, exc, "any", limit=1, file=log, chain=True)
   # Output is same in all the cases.
```

**Comment by microDev1** [comment]
> Is this parameter required in CPython?

Yes, this is a required parameter.

> What will we provide in CircuitPython? 

As this is ignored, anything (i.e. `Type: Any`) can be supplied to the function call.

> Does it do the same thing in CPython?

Yes, the current implementation is same as in `CPython`. The only difference is:

- `CPython` - The traceback is obtained from the supplied traceback object.
- `CircuitPython` - There is no traceback object (i.e. `Type: TracebackType`) implemented so this is obtained internally.

Only thing I can think of that won't be possible because of the above is supplying traceback from a different exception.

> Would you mind providing an example use?

```python
try:
    raise Exception("xyz")
except Exception as exc:
    # CPython:
    traceback.print_exception(None, exc, None)
    traceback.print_exception(None, exc, exc.__traceback__)
    traceback.print_exception(None, exc, exc.__traceback__, limit=1, file=log, chain=True)
    # CircuitPython:
    traceback.print_exception(None, exc, None)
    traceback.print_exception(None, exc, "any")
    traceback.print_exception(None, exc, "any", limit=1, file=log, chain=True)
    # Output is same in all the cases to there CPython counterpart.
```

**Comment by microDev1** [comment]
I can enable `__traceback__` and make it return `0` for now, this will make function calls exactly the same.

**Comment by tannewt** [comment]
Why don't we make an empty Traceback object type and every exception can have its own. That way we can verify that the given `__traceback__` is from the provided exception. I want to avoid cases where folks can provide values that are invalid in CPython.

File: shared-bindings/traceback/__init__.c [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+//|     .. note: Setting `chain` will have no effect as chained exceptions are not yet implemented.
+//|
+//|     :param Type[BaseException] etype: This is ignored and inferred from the type of ``value``.
+//|     :param BaseException value: The exception. Must be an instance of `BaseException`.
+//|     :param TracebackType tb: The traceback object. If `None`, the traceback will not be printed.
+//|     :param int limit: Print up to limit stack trace entries (starting from the caller’s frame) if limit is positive.
+//|                       Otherwise, print the last ``abs(limit)`` entries. If limit is omitted or None, all entries are printed.
+//|     :param io.FileIO file: If file is omitted or `None`, the output goes to `sys.stderr`; otherwise it should be an open
+//|                            file or file-like object to receive the output.
+//|     :param bool chain: If `True` then chained exceptions will be printed.
```
Comment (tannewt) [inline shared-bindings/traceback/__init__.c:53]: I'd just comment this here that it isn't implemented.

File: shared-bindings/traceback/__init__.c [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+    mp_int_t limit = 0;
+    bool print_tb = true;
+    if (args[ARG_limit].u_obj != mp_const_none) {
+        if (!mp_obj_get_int_maybe(args[ARG_limit].u_obj, &limit)) {
+            mp_raise_TypeError(translate("limit should be an int"));
+        }
+        print_tb = !(limit == 0);
+    }
+
```
Comment (tannewt) [inline shared-bindings/traceback/__init__.c:96]: Please put everything below in shared-module. It's easier to read shared bindings if it only does argument validation.

**Review by tannewt — CHANGES_REQUESTED** [review]
Just a couple of things. This is getting closer and closer. Thanks for taking it on!

**Review by tannewt — APPROVED** [review]
Thank you!

**Pull request merged**